File: server/cpp_server/src/dag/node.cc
/*
 * Copyright 2022 netease. All rights reserved.
 * Author zhaochaochao@corp.netease.com
 * Date   2023/02/15
 * Brief  Node of DAG.
 */

#include "node.h"

#include <butil/time.h>

#include "logger/logger.h"

namespace netease::grps {
void ModelNode::Process(const ::grps::protos::v1::GrpsMessage& input,
                        ::grps::protos::v1::GrpsMessage& output,
                        netease::grps::GrpsContext& ctx) {
  auto begin = butil::gettimeofday_us();
  ctx.set_converter(converter_.get());
  ctx.set_inferer(model_inferer_.get());

  if (batcher_) {
    batcher_->Infer(input, output, ctx);
    return;
  }

  if (converter_) {
    std::vector<std::pair<std::string, TensorWrapper>> inp_tensors;
    std::vector<std::pair<std::string, TensorWrapper>> out_tensors;

    converter_->PreProcess(input, inp_tensors, ctx);
    if (ctx.has_err()) {
      return;
    }
    auto preprocess_end = butil::gettimeofday_us();

    model_inferer_->Infer(inp_tensors, out_tensors, ctx);
    if (ctx.has_err()) {
      return;
    }
    auto infer_end = butil::gettimeofday_us();

    output.Clear();
    converter_->PostProcess(out_tensors, output, ctx);
    if (ctx.has_err()) {
      return;
    }
    auto postprocess_end = butil::gettimeofday_us();

    LOG4(INFO, "Model(" << name_ << "), preprocess latency: " << preprocess_end - begin
                        << "us, infer latency: " << infer_end - preprocess_end
                        << "us, postprocess latency: " << postprocess_end - infer_end << "us");
  } else {
    output.Clear();
    model_inferer_->Infer(input, output, ctx);
    if (ctx.has_err()) {
      return;
    }
    auto infer_end = butil::gettimeofday_us();
    LOG4(INFO, "Model(" << name_ << "), infer latency: " << infer_end - begin << "us");
  }
}

void ModelNode::Process(const ::grps::protos::v1::GrpsMessage& input,
                        ::grps::protos::v1::GrpsMessage& output,
                        const std::shared_ptr<GrpsContext>& ctx_sp) {
  auto begin = butil::gettimeofday_us();
  ctx_sp->set_converter(converter_.get());
  ctx_sp->set_inferer(model_inferer_.get());

  if (batcher_) {
    batcher_->Infer(input, output, ctx_sp);
    return;
  }

  if (converter_) {
    std::vector<std::pair<std::string, TensorWrapper>> inp_tensors;
    std::vector<std::pair<std::string, TensorWrapper>> out_tensors;

    converter_->PreProcess(input, inp_tensors, *ctx_sp);
    if (ctx_sp->has_err()) {
      return;
    }
    auto preprocess_end = butil::gettimeofday_us();

    model_inferer_->Infer(inp_tensors, out_tensors, *ctx_sp);
    if (ctx_sp->has_err()) {
      return;
    }
    auto infer_end = butil::gettimeofday_us();

    output.Clear();
    converter_->PostProcess(out_tensors, output, *ctx_sp);
    if (ctx_sp->has_err()) {
      return;
    }
    auto postprocess_end = butil::gettimeofday_us();

    LOG4(INFO, "Model(" << name_ << "), preprocess latency: " << preprocess_end - begin
                        << "us, infer latency: " << infer_end - preprocess_end
                        << "us, postprocess latency: " << postprocess_end - infer_end << "us");
  } else {
    output.Clear();
    model_inferer_->Infer(input, output, *ctx_sp);
    if (ctx_sp->has_err()) {
      return;
    }
    auto infer_end = butil::gettimeofday_us();
    LOG4(INFO, "Model(" << name_ << "), infer latency: " << infer_end - begin << "us");
  }
}

void ModelNode::Process(const std::vector<::grps::protos::v1::GrpsMessage>& input,
                        std::vector<::grps::protos::v1::GrpsMessage>& output,
                        GrpsContext& ctx) {}

void ModelNode::Process(const std::vector<::grps::protos::v1::GrpsMessage>& input,
                        std::vector<::grps::protos::v1::GrpsMessage>& output,
                        const std::shared_ptr<GrpsContext>& ctx_sp) {}

void MergerNode::Process(const ::grps::protos::v1::GrpsMessage& input,
                         ::grps::protos::v1::GrpsMessage& output,
                         GrpsContext& ctx) {}

void MergerNode::Process(const ::grps::protos::v1::GrpsMessage& input,
                         ::grps::protos::v1::GrpsMessage& output,
                         const std::shared_ptr<GrpsContext>& ctx_sp) {}

void MergerNode::Process(const std::vector<::grps::protos::v1::GrpsMessage>& input,
                         std::vector<::grps::protos::v1::GrpsMessage>& output,
                         GrpsContext& ctx) {}

void MergerNode::Process(const std::vector<::grps::protos::v1::GrpsMessage>& input,
                         std::vector<::grps::protos::v1::GrpsMessage>& output,
                         const std::shared_ptr<GrpsContext>& ctx_sp) {}

void SplitterNode::Process(const ::grps::protos::v1::GrpsMessage& input,
                           ::grps::protos::v1::GrpsMessage& output,
                           GrpsContext& ctx) {}

void SplitterNode::Process(const ::grps::protos::v1::GrpsMessage& input,
                           ::grps::protos::v1::GrpsMessage& output,
                           const std::shared_ptr<GrpsContext>& ctx_sp) {}

void SplitterNode::Process(const std::vector<::grps::protos::v1::GrpsMessage>& input,
                           std::vector<::grps::protos::v1::GrpsMessage>& output,
                           GrpsContext& ctx) {}

void SplitterNode::Process(const std::vector<::grps::protos::v1::GrpsMessage>& input,
                           std::vector<::grps::protos::v1::GrpsMessage>& output,
                           const std::shared_ptr<GrpsContext>& ctx_sp) {}
} // namespace netease::grps

File: server/py_server/grps_framework/dag/node.py
# Copyright 2022 netease. All rights reserved.
# Author zhaochaochao@corp.netease.com
# Date   2022/9/14
# Brief  Node of DAG.
import time
from abc import ABCMeta, abstractmethod

from grps_framework.context.context import GrpsContext
from grps_framework.logger.logger import logger


class NodeBase(metaclass=ABCMeta):
    """Node base class."""

    @abstractmethod
    def __init__(self, name):
        """Initiate node base class."""
        self._name = name

    @abstractmethod
    def process(self, data, context):
        """
        Node process.
        Args:
            context: grps context
            data: node input.

        Returns:
            node process output.
        """
        pass

    @property
    def name(self):
        return self._name

    def __str__(self):
        return self._name


class ModelNode(NodeBase):
    """Model node, which means the node is a model."""

    def __init__(self, name, model_infer, converter, batcher):
        """
        Init model node.
        Args:
            name: node name.
            model_infer: Model infer instance.
            converter: Converter instance.
            batcher: Batcher instance.
        """
        super(ModelNode, self).__init__(name)
        self.__model_infer = model_infer
        self.__converter = converter
        self.__batcher = batcher

    def process(self, data, context: GrpsContext):
        """
        Node process.
        Args:
            context: grps context
            data: node input.

        Returns:
            node process output.
        """
        begin = time.time()

        context.set_converter(self.__converter)
        context.set_inferer(self.__model_infer)

        if self.__batcher is not None:
            return self.__batcher.infer(data, context)

        if self.__converter is None:
            infer_in = data
            infer_out = self.__model_infer.infer(infer_in, context)
            # logger.debug('Infer_input: {}, infer_out: {}'.format(infer_input, infer_out))
            infer_time = time.time()
            if context.has_err():
                return

            logger.info(
                'Model({}), model_infer time: {:.0f} us'.format(self._name, (infer_time - begin) * 1e6))
            return infer_out
        else:
            infer_input = self.__converter.preprocess(data, context)
            preprocess_time = time.time()
            if context.has_err():
                return

            infer_out = self.__model_infer.infer(infer_input, context)
            # logger.debug('Infer_input: {}, infer_out: {}'.format(infer_input, infer_out))
            infer_time = time.time()
            if context.has_err():
                return

            out = self.__converter.postprocess(infer_out, context)
            postprocess_time = time.time()
            if context.has_err():
                return

            logger.info(
                'Model({}), preprocess time: {:.0f} us, model_infer time: {:.0f} us, postprocess time: {:.0f} us'.format(
                    self._name, (preprocess_time - begin) * 1e6, (infer_time - preprocess_time) * 1e6,
                                (postprocess_time - infer_time) * 1e6))

            return out


# TODO: Add splitter node.
class SplitterNode(NodeBase):
    """Splitter node, which means the node is a splitter."""

    def __init__(self, name):
        pass

    def process(self, data, context):
        pass


# TODO: Add merger node.
class MergerNode(NodeBase):
    """Merger node, which means the node is a merger."""

    def __init__(self, name):
        pass

    def process(self, data, context):
        pass
